odoo/addons/base/models/res_device.py

```diff
@@ -0,0 +1,184 @@
+
+            :param request: Request or WebsocketRequest object
+        """
+        user_agent = request.httprequest.user_agent
+        platform = user_agent.platform
+        browser = user_agent.browser
+        ip_address = request.httprequest.remote_addr
+
+        trace = request.session.update_trace(platform, browser, ip_address)
+        if not trace:
+            return
+
+        user_id = request.session.uid
+        geoip = GeoIP(ip_address)
+        session_identifier = request.session.sid[:42]
+
+        if self.env.cr.readonly:
+            cursor = self.env.registry.cursor(readonly=False)
+        else:
+            cursor = nullcontext(self.env.cr)
+        with cursor as cr:
+            cr.execute(SQL("""
+                INSERT INTO res_device_log (session_identifier, platform, browser, ip_address, country, city, device_type, user_id, first_activity, last_activity, revoked)
+                VALUES (%(session_identifier)s, %(platform)s, %(browser)s, %(ip_address)s, %(country)s, %(city)s, %(device_type)s, %(user_id)s, %(first_activity)s, %(last_activity)s, %(revoked)s)
+            """,
+                session_identifier=session_identifier,
+                platform=platform,
+                browser=browser,
+                ip_address=ip_address,
+                country=geoip.get('country_name'),
+                city=geoip.get('city'),
+                device_type='mobile' if self._is_mobile(platform) else 'computer',
+                user_id=user_id,
+                first_activity=datetime.fromtimestamp(trace['first_activity']),
+                last_activity=datetime.fromtimestamp(trace['last_activity']),
+                revoked=False,
+            ))
+        _logger.info("User %d inserts device log (%s)", user_id, session_identifier)
+
+    @api.autovacuum
+    def _gc_device_log(self):
+        # Keep the last device log
+        # (even if the session file no longer exists on the filesystem)
+        self.env.cr.execute("""
+            DELETE FROM res_device_log log1
+            WHERE EXISTS (
+                SELECT 1 FROM res_device_log log2
+                WHERE
+                    log1.session_identifier = log2.session_identifier
+                    AND log1.platform = log2.platform
+                    AND log1.browser = log2.browser
+                    AND log1.ip_address = log2.ip_address
+                    AND log1.last_activity < log2.last_activity
+            )
+        """)
+        _logger.info("GC device logs delete %d entries", self.env.cr.rowcount)
+
+
+class ResDevice(models.Model):
+    _name = "res.device"
+    _inherit = ["res.device.log"]
+    _description = "Devices"
+    _auto = False
+
+    @check_identity
+    def revoke(self):
+        return self._revoke()
+
+    def _revoke(self):
+        ResDeviceLog = self.env['res.device.log']
+        session_identifiers = list({device.session_identifier for device in self})
```

Comment:
Nitpicking: Not really important but it is not deterministic to use a set because of the PYTHONHASHSEED. => `list(unique(device.session_identifier for device in self))`